extension/source/QuillPage/components/PasswordCreationForm.tsx

```diff
@@ -1,27 +1,113 @@
 import * as React from 'react';
+
+import measurePasswordStrength, {
+  PasswordStrength,
+} from '../../helpers/measurePasswordStrength';
 import QuickColumn from './QuickColumn';
 import QuickRow from './QuickRow';
 
-const PasswordCreationForm: React.FunctionComponent = () => (
-  <div className="password-creation-form quick-column">
-    <QuickColumn>
-      <input
-        type="password"
-        placeholder="Password"
-        style={{ width: '100%', flexGrow: 0 }}
-      />
-      <input
-        type="password"
-        placeholder="Confirm password"
-        style={{ width: '100%', flexGrow: 0 }}
-      />
-      <QuickRow>
-        <div>Password strength</div>
-        <div>Average</div>
-      </QuickRow>
-      <div>This password is ok, but could be more secure.</div>
-    </QuickColumn>
-  </div>
-);
+const passwordCommentaryMap: Record<PasswordStrength['descriptor'], string> = {
+  'Very weak': 'Please consider using a stronger password.',
+  Weak: 'Please consider using a stronger password.',
+  Average:
+    "This password is about average, which isn't great for protecting assets.",
+  Good: 'This is a relatively good password, but could be more secure.',
+  Strong: '',
+};
+
+const PasswordCreationForm: React.FunctionComponent<{
+  onPasswordUpdate: (password: string | undefined) => void;
+}> = ({ onPasswordUpdate }) => {
+  const [password, setPassword] = React.useState<string>();
+
+  const [passwordFieldValue, setPasswordFieldValue] = React.useState('');
+  const [confirmPasswordFieldValue, setConfirmPasswordFieldValue] =
+    React.useState('');
+
+  function handleFieldsChange(newPassword: string, newConfirmPassword: string) {
+    if (newPassword !== passwordFieldValue) {
+      setPasswordFieldValue(newPassword);
+    }
+
+    if (newConfirmPassword !== confirmPasswordFieldValue) {
+      setConfirmPasswordFieldValue(newConfirmPassword);
+    }
+
+    const newResultPassword =
+      newPassword === newConfirmPassword ? newPassword : undefined;
+
+    if (newResultPassword !== password) {
+      setPassword(newResultPassword);
+      onPasswordUpdate(newResultPassword);
+    }
+  }
+
+  const passwordStrength = measurePasswordStrength(passwordFieldValue);
+
+  const confirmFieldClasses = ['password-confirm-field'];
+
+  if (confirmPasswordFieldValue === '') {
+    confirmFieldClasses.push('empty');
+  } else if (confirmPasswordFieldValue === passwordFieldValue) {
+    confirmFieldClasses.push('correct');
+  } else if (passwordFieldValue.startsWith(confirmPasswordFieldValue)) {
+    confirmFieldClasses.push('incomplete');
+  } else {
+    confirmFieldClasses.push('incorrect');
+  }
```

Comment:
To prevent redundant computation in render cycles and prevent subcomponents from unnecessarily re-rendering, generating values like this can be memoized. https://reactjs.org/docs/hooks-reference.html#usememo

```suggestion
  const confirmFieldClasses = React.useMemo(() => {
    const classes = ['password-confirm-field'];
    if (confirmPasswordFieldValue === '') {
      classes.push('empty');
    } else if (confirmPasswordFieldValue === passwordFieldValue) {
      classes.push('correct');
    } else if (passwordFieldValue.startsWith(confirmPasswordFieldValue)) {
      classes.push('incomplete');
    } else {
      classes.push('incorrect');
    }
    return classes;
  }, [confirmPasswordFieldValue, passwordFieldValue]);
```
Now `confirmFieldClasses` will only become a new `Array` object when either `confirmPasswordFieldValue` or `passwordFieldValue` changes.

I don't think this is needed for this round of changes, but something to keep in mind.